presto-main/src/main/java/io/prestosql/sql/planner/PlanOptimizers.java

```diff
@@ -251,23 +300,20 @@ public void destroy()
         optimizerStats.unexport(exporter);
     }
 
-    public PlanOptimizers(
-            Metadata metadata,
-            TypeAnalyzer typeAnalyzer,
-            TaskManagerConfig taskManagerConfig,
-            boolean forceSingleNode,
-            MBeanExporter exporter,
-            SplitManager splitManager,
-            PageSourceManager pageSourceManager,
-            StatsCalculator statsCalculator,
-            CostCalculator costCalculator,
-            CostCalculator estimatedExchangesCostCalculator,
-            CostComparator costComparator,
-            TaskCountEstimator taskCountEstimator)
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        return collectOptimizers(new OptimizingBuilder(plan, session, symbolAllocator, idAllocator, warningCollector));
+    }
+
+    @VisibleForTesting
+    public List<PlanOptimizer> getOptimizers()
     {
-        this.exporter = exporter;
-        ImmutableList.Builder<PlanOptimizer> builder = ImmutableList.builder();
+        return collectOptimizers(new CollectingBuilder());
+    }
 
+    private <T> T collectOptimizers(OptimizationBuilder<T> builder)
```

Comment:
This means optimizers are going to be instantiated over and over for each query. That can have non-trivial cost on the kind work the coordinator does. Also, it's a departure from the current model where optimizers are instantiated once with the query-agnostic dependencies (metadata, etc) and have the per-query elements (session, warning collector, etc) provided during the optimize call. This introduces a mode that is a combination of both and confuses things a bit.

Reply:
That's correct. Also, that seems inherent. The optimizers' contructors do not do much work, but there is eg rule indexing.
Likely this can be dominant factor for very small queries.